yarn-project/sequencer-client/src/slasher/slasher_client.ts

```diff
@@ -0,0 +1,408 @@
+    return this.synchedProvenBlockNumber.get() ?? INITIAL_L2_BLOCK_NUM - 1;
+  }
+
+  /**
+   * Method to check the status of the slasher client.
+   * @returns Information about slasher client status: state & syncedToBlockNum.
+   */
+  public async getStatus(): Promise<SlasherSyncState> {
+    const blockNumber = this.getSyncedLatestBlockNum();
+    const blockHash =
+      blockNumber == 0
+        ? ''
+        : await this.l2BlockSource.getBlockHeader(blockNumber).then(header => header?.hash().toString());
+    return Promise.resolve({
+      state: this.currentState,
+      syncedToL2Block: { number: blockNumber, hash: blockHash },
+    } as SlasherSyncState);
+  }
+
+  /**
+   * Handles new mined blocks by marking the txs in them as mined.
+   * @param blocks - A list of existing blocks with txs that the slasher client needs to ensure the tx pool is reconciled with.
+   * @returns Empty promise.
+   */
+  private async handleLatestL2Blocks(blocks: L2Block[]): Promise<void> {
+    if (!blocks.length) {
+      return Promise.resolve();
+    }
+
+    const lastBlockNum = blocks[blocks.length - 1].number;
+    await Promise.all(blocks.map(block => this.synchedBlockHashes.set(block.number, block.hash().toString())));
+    await Promise.all(
+      blocks.map(block => this.synchedBlockSlots.set(block.number, block.header.globalVariables.slotNumber.toBigInt())),
+    );
+    await this.synchedLatestBlockNumber.set(lastBlockNum);
+    this.log.debug(`Synched to latest block ${lastBlockNum}`);
+    this.startServiceIfSynched();
+  }
+
+  /**
+   * Handles new proven blocks by deleting the txs in them, or by deleting the txs in blocks `keepProvenTxsFor` ago.
+   * @param blocks - A list of proven L2 blocks.
+   * @returns Empty promise.
+   */
+  private async handleProvenL2Blocks(blocks: L2Block[]): Promise<void> {
+    if (!blocks.length) {
+      return Promise.resolve();
+    }
+    const lastBlockNum = blocks[blocks.length - 1].number;
+    await this.synchedProvenBlockNumber.set(lastBlockNum);
+    this.log.debug(`Synched to proven block ${lastBlockNum}`);
+
+    this.startServiceIfSynched();
+  }
+
+  /**
+   * Updates the tx pool after a chain prune.
+   * @param latestBlock - The block number the chain was pruned to.
+   */
+  private async handlePruneL2Blocks(latestBlock: number): Promise<void> {
+    const slotNumber = this.synchedBlockSlots.get(latestBlock) ?? BigInt(0);
+    const epochNumber = slotNumber / BigInt(this.config.aztecEpochDuration);
+    this.log.info(`Detected chain prune. Punishing the validators at epoch ${epochNumber}`);
+
+    // Set the lifetime such that we have a full round that we could vote throughout.
+    const slotsIntoRound = slotNumber % BigInt(this.config.slashingRoundSize);
+    const toNext = slotsIntoRound == 0n ? 0n : BigInt(this.config.slashingRoundSize) - slotsIntoRound;
+
+    const lifetime = slotNumber + toNext + BigInt(this.config.slashingRoundSize);
+
+    this.slashEvents.push({
```

Comment:
another thing about this design, does every voting client need to detect the offense themselves to be able to vote on it ?

Reply:
In the current impl yes. So not all nodes would see the same, it would be based on when they are online.
But it is as stated in the beginning in there  only a very brute implementation. I think a better "real" implementation would be looking for events for this specific kind of action or accepting "evidence" from other parties as well. This is mainly to put in some of the plumbing and show that it can be done.